backend/dataall/modules/notebooks/aws/client.py

```diff
@@ -0,0 +1,62 @@
+import logging
+
+from dataall.aws.handlers.sts import SessionHelper
+from dataall.modules.notebooks.db.models import SagemakerNotebook
+from botocore.exceptions import ClientError
+
+logger = logging.getLogger(__name__)
+
+
+class SagemakerClient:
+    """
+    A Sagemaker notebooks proxy client that is used to send requests to AWS
+    """
+
+    def __init__(self, notebook: SagemakerNotebook):
```

Comment:
Did you think about organising it in a bit different way? 
Seems like we will be creating a separate SageMaker client per the SageMaker notebook instance individually but the clients are rather aws account id, region specific. So if there are requests for notebooks belonging to the same awsAccount, region not sure if it will be the most optimal solution to create a new client for each of these?

I see we are passing the full notebook objects to the constructor but actually only the accountId, region, instance_name are used.

I would rather think of creating a client per aws account id/region and use a flyweight pattern to provide notebook or instance name to the methods for getting a status etc.

Reply:
I don't see a big difference of two solutions from the technical perspective. But in your provided example, we will do something like this in every single line when we want to use:
```
client(notebook.AWSAccountId, notebook.region).start_instance(notebook.NotebookInstanceName)
```
That will make us to write more code.  But let's also assume what happens if we need more parameters in the future? For example, we want to pass VPC, subnets, and InstanceType. The number of parameters will keep growing.

About the caching: this code will be executed in lambda, so the creation of the local cache with future re-using is not possible. 

For me, the original version has a better meaning: it's a AWS client for a specific Sagemaker notebook. You just need to pass the instance and the client will encapsulate the logic how the request to AWS will be executed. 
 
Of course, it can be revisited later when we see how the other `clients` will be organize. 🙂

Reply:
I think like we don't need to create a new client for every request/a need to execute a sagemaker api operation.
For lambda the best practices the client should be created before the handler is executed and if the same lambda container will be reused for the next call we can reuse the client (for same region,account id) otherwise we are increasing the request handling time by creating a new client every time. 

 But I guess it is not the most important for now so I agree that we can get back to it at some point.

Reply:
but we indeed here have a bit of a challenge since we don't know the account id / region before the request will be sent so let's revisit it later

Reply:
An alternative is to cache the clients per account/region e.g.:

```
from functools import cache

@cache
def _get_client(account, region):
    session = ...
    return session.client(...)
```

Reply:
I don't think that the cache is needed here, since there is a huge chance that lambda execution will be finished right after the request. Since Lambda is stateless in the nature, after adding a cache, we might end up in the situation where we cached a client, the lambda process finished with the request, and a new lambda will start on the next request without the cached client.
But I like this cache from functools :)